devicetypes/smartthings/zigbee-accessory-dimmer.src/zigbee-accessory-dimmer.groovy

```diff
@@ -43,64 +44,160 @@ metadata {
+			results << createEvent(name: "switch", value: device.currentValue("switch") == "on" ? "off" : "on")
+		} else {
+			results << createEvent(name: "switch", value: descMap.commandInt == ONOFF_OFF_COMMAND ? "off" : "on")
+		}
+	}
+
+	return results
+}
+
+def handleLevelEvent(descMap) {
+	def results = []
+
+	if (descMap.commandInt == LEVEL_STEP_COMMAND) {
+		results += handleStepEvent(descMap)
+	} else if (descMap.commandInt == LEVEL_MOVE_COMMAND || descMap.commandInt == LEVEL_MOVE_ONOFF_COMMAND) {
+		// For the IKEA Dimmer we are going to treat Level Move and Level Move with On/Off as Level Step
+		if (isIkeaDimmer()) {
+			results += handleStepEvent(descMap)
+		} else {
+			def newLevel = descMap.data[0] == LEVEL_DIRECTION_UP ? 100 : STEP
+			results << createEvent(name: "level", value: newLevel)
+			results << createEvent(name: "switch", value: "on")
+
+			log.debug "step $STEP to ${newLevel}"
+		}
+	} else if (descMap.commandInt == LEVEL_STOP_COMMAND || descMap.commandInt == LEVEL_STOP_ONOFF_COMMAND) {
+		// We are not going to handle this event because we are not implementing this the way that the Zigbee spec indicates
+		log.debug "Received stop move - not handling"
+	} else if (descMap.commandInt == LEVEL_MOVE_LEVEL_ONOFF_COMMAND) {
+		// The spec defines this as "Move to level with on/off". The IKEA Dimmer sends us 0x00 or 0xFF only, so we will treat this more as a
+		// on/off command for the dimmer. Otherwise, we will treat this as off or on and setLevel.
+		if (descMap.data[0] == "00") {
+			results << createEvent(name: "switch", value: "off", isStateChange: true)
+		} else if (descMap.data[0] == "FF" && isIkeaDimmer()) {
+			// The IKEA Dimmer sends 0xFF -- this is technically not to spec, but we will treat this as an "on"
+			if (device.currentValue("level") == 0) {
+				results << createEvent(name: "level", value: STEP)
+			}
+
+			results << createEvent(name: "switch", value: "on", isStateChange: true)
+		} else {
+			results << createEvent(name: "switch", value: "on", isStateChange: true)
+			// Handle the Zigbee level the same way as we would normally with the same code path -- commandInt doesn't matter right now
+			// The first byte is the level, the second two bytes are the rate -- we only care about the level right now.
+			results << createEvent(zigbee.getEventFromAttrData(descMap.clusterInt, descMap.commandInt, UINT8_STR, descMap.data[0]))
+		}
+	}
+
+	return results
+}
+
+def handleStepEvent(descMap) {
+	def results = []
+	def currentLevel = device.currentValue("level") as Integer ?: 0
+	def value = null
+
+	log.debug "handleStepEvent - ${descMap.data}"
+
+	if (descMap.data[0] == LEVEL_DIRECTION_UP) {
+		value = Math.min(currentLevel + STEP, 100)
+	} else if (descMap.data[0] == LEVEL_DIRECTION_DOWN) {
+		value = Math.max(currentLevel - STEP, 0)
+	}
+
+	if (value != null) {
+		log.debug "Step ${descMap.data[0] == LEVEL_DIRECTION_UP ? "up" : "down"} by $STEP to $value"
```

Comment:
Could move this up to the if/else if and then you'd know it was up or down already